src/record_ts.h

```diff
@@ -0,0 +1,629 @@
+#include "drjit-core/hash.h"
+#include "drjit-core/jit.h"
+#include "internal.h"
+#include <cstdint>
+
+void jitc_freeze_start(JitBackend backend, const uint32_t *inputs,
+                       uint32_t n_inputs);
+
+Recording *jitc_freeze_stop(JitBackend backend, const uint32_t *outputs,
+                            uint32_t n_outputs);
+
+void jitc_freeze_abort(JitBackend backend);
+
+void jitc_freeze_destroy(Recording *recording);
+
+int jitc_freeze_pause(JitBackend backend);
+
+int jitc_freeze_resume(JitBackend backend);
+
+void jitc_freeze_replay(Recording *recording, const uint32_t *inputs,
+                        uint32_t *outputs);
+
+int jitc_freeze_dry_run(Recording *recording, const uint32_t *inputs);
+
+/// HashMap, mapping an allocation to a recorded variable
+using PtrToSlot = tsl::robin_map<const void *, uint32_t, PointerHasher>;
+
+enum class OpType {
+    Barrier,
+    KernelLaunch,
+    MemsetAsync,
+    Expand,
+    ReduceExpanded,
+    Compress,
+    MemcpyAsync,
+    Mkperm,
+    BlockReduce,
+    BlockPrefixReduce,
+    ReduceDot,
+    Aggregate,
+    Free,
+    Count,
+};
+
+extern const char *op_type_name[(int) OpType::Count];
+
+/**
+ * Represents an operation that was recorded by a \ref RecordThreadState. To
+ * record and operation, we refer to the `reduce_dot` example at the top of this
```

Comment:
There isn't such an example in this file, perhaps you meant to reference the `.cpp` file?

Reply:
Ah right, I changed that before moving the top comment to the `.cpp` file. I'll adjust it.